### PR TITLE
EVG-2124: run host init in threadpool

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: hostinit/setup.go [hunk 1]
```diff
@@ -137,83 +137,88 @@ func (init *HostInit) setupReadyHosts(ctx context.Context) error {
...
-			grip.Debug(message.Fields{
-				"GUID":    init.GUID,
-				"message": "host not ready for setup",
-				"hostid":  h.Id,
-				"DNS":     h.Host,
-			})
-			continue
-		}
+	var numThreads int
+	if len(hosts) <= 16 {
```
Comment (john-m-liu) [inline hostinit/setup.go]: should be >=?

File: hostinit/setup.go [hunk 2]
```diff
@@ -137,83 +137,88 @@ func (init *HostInit) setupReadyHosts(ctx context.Context) error {
 
 	// used for making sure we don't exit before a setup script is done
 	wg := &sync.WaitGroup{}
+	hosts := make(chan host.Hosts, len(uninitializedHosts))
```
Comment (john-m-liu) [inline hostinit/setup.go]: should be host.Host?

File: hostinit/setup.go [hunk 3]
```diff
@@ -137,83 +137,88 @@ func (init *HostInit) setupReadyHosts(ctx context.Context) error {
...
-				hostLink := fmt.Sprintf("%v/host/%v", init.Settings.Ui.Url, h.Id)
-				message := fmt.Sprintf("Provisioning failed on %v host -- %v: see %v",
-					h.Distro.Id, h.Id, hostLink)
-				if err := notify.NotifyAdmins(subject, message, init.Settings); err != nil {
-					grip.Errorf("Error sending email: %+v", err)
+		go func() {
+			defer wg.Done()
+			for h := range hosts {
+				if ctx.Err() != nil {
+					return errors.New("hostinit run canceled")
```
Comment (john-m-liu) [inline hostinit/setup.go]: the goroutine doesn't return anything

File: hostinit/setup.go [hunk 4]
```diff
@@ -137,83 +137,94 @@ func (init *HostInit) setupReadyHosts(ctx context.Context) error {
...
+				if err != nil {
+					err = errors.Wrapf(err, "problem checking host %s for readiness", h.Id)
+					catcher.Add(err)
+					grip.Error(err.Error())
+					continue
+				}
 
-			wg.Done()
+				// if the host isn't ready (for instance, it might not be up yet), skip it
+				if !ready {
```
Comment (bsamek) [inline hostinit/setup.go]: log the time between this and line 164

File: hostinit/setup.go [hunk 5]
```diff
@@ -137,83 +137,94 @@ func (init *HostInit) setupReadyHosts(ctx context.Context) error {
...
-				// notify the admins of the failure
-				subject := fmt.Sprintf("%v Evergreen provisioning failure on %v",
-					notify.ProvisionFailurePreface, h.Distro.Id)
-				hostLink := fmt.Sprintf("%v/host/%v", init.Settings.Ui.Url, h.Id)
-				message := fmt.Sprintf("Provisioning failed on %v host -- %v: see %v",
-					h.Distro.Id, h.Id, hostLink)
-				if err := notify.NotifyAdmins(subject, message, init.Settings); err != nil {
-					grip.Errorf("Error sending email: %+v", err)
+		go func() {
+			defer wg.Done()
```
Comment (bsamek) [inline hostinit/setup.go]: Where do we wait?

Comment (tychoish) [reply hostinit/setup.go]: ln 232

**Pull request merged**